### DISCUSSION TITLE
docker-compose单机版2.2.9，buff/cache增长快

### DISCUSSION THREAD

**Discussion by ruixiu** [milvus-io/milvus#24816] [Q&A and General discussion]
我们的应用场景无法使用批量插入，所以在逐条且频繁插入数据的时候，很快出现内存不足的情况。使用milvus早期版本的时候，观察发现实际占用内存并不高，但是buff/cache占用呈持续增长状态。根据查到的资料，我们做了定时检查可用内存，在可用内存不足时释放buff/cache的脚本，临时解决了。之后部署milvus一直使用的docker-compose运行集群版，没再出现过因内存问题导致不可用的情况。
直至升级到2.2.8（仍然使用的docker-compose运行的集群版），频繁出现磁盘占用爆满的情况，重启docker容器后即释放。根据官方建议升级到2.2.9（使用docker-compose运行的单机版），再次出现了buff/cache持续增长的问题，并且更严重的是此时占用的磁盘存储空间也呈持续增长状态，经观察发现docker容器的overlay2占用磁盘空间并不多。很无解，不知道应该如何使用或处理。【客观条件限制了无法使用k8s部署集群版】
![微信图片_20230612110509](https://github.com/milvus-io/milvus/assets/9940354/c9e099a5-bb59-4e3d-a395-9836cfbc8df6)
![微信图片_20230612110517](https://github.com/milvus-io/milvus/assets/9940354/337de66f-7276-4f2e-ab37-59c3abe42895)

**Comment by yhmo** [comment]
能否告知你们的客户端插入数据的具体行为，逐条插入是多个客户端在做么？每秒大概会有多少条进入milvus？是否调用了flush()? 是否插入同一个collection？是否有多个partition？除了向量字段之外，有没有其他字符串字段？milvus.yaml的配置有些什么改动？

**Reply by ruixiu** [reply]
1.只有一个客户端，具体是这样的：数据采集程序持续采集数据送到服务端，服务端提取特征向量后在milvus中查询一次找到最相似的一个，如果相似度阈值低于85%就执行向量插入操作
2.每秒在10条之内，没有明确调用flush()
3.我们是针对车头的相似度查询场景，共有2个collection，一个全车头collection一次插入1条，一个分区域collection循环插入9条
4.所有collection都只有默认partition（最开始的时候尝试过使用partition，但是内存占用过高，我们就放弃分partition了）
5.每个collection除了向量字段，还有8个标量字段（包含主键id是自增的）
6.没有改动milvus.yaml，全部都是默认的

**Reply by yhmo** [reply]
每秒10条不会产生很大的数据量，milvus真正占用的内存不会很多。buf/cache是由系统管理的，我理解是无论手动释放还是系统自动释放都不会影响milvus的内存使用。

**Comment by xiaofan-luan** [comment]
@LoveEachDay 
any comments on this issue?
Seems to be related to data cleanup on docker?

**Comment by LoveEachDay** [comment]
@ruixiu 
1. For the memory increasement for `buf/cache`, this is normal for linux os, which is also usable for applications. You can learn more from: https://www.linuxatemyram.com/
2. For the container disk usage, the part which uses more than 100GB is `merged` directory, which means the container write many files, could you help us to identify which container has such a disk usage.

**Reply by ruixiu** [reply]
docker容器占用的并不多，但是磁盘占满了，所以怀疑是不是因为buff/cache持续增长导致的（但是没有判断依据）
![image](https://github.com/milvus-io/milvus/assets/9940354/9059d486-2b5d-4f8d-a54d-eaa3c9682bf1)
另外，我对系统盘使用du -sh检查了一下，实际占用也不多，如下图：【我们的/workspace/data是挂载的另外一个磁盘，用来存储所有数据的】
![image](https://github.com/milvus-io/milvus/assets/9940354/4493aef1-8611-41a9-b962-91501d221629)

**Reply by yhmo** [reply]
/var/lib/docker/overlay2 占用300G很可疑，可以查一下是什么文件占用了那么大空间。类似这个：https://blog.csdn.net/chenfengkai/article/details/122436707

如果是日志的问题，可以把日志级别改一下，在milvus.yaml里的log.level设置

**Reply by ruixiu** [reply]
查了，查不出来到底哪个文件占用的。

**Comment by ruixiu** [comment]
占用300G是因为centos-root只有300G。
原来是100G，满了后以为数据量大确实需要大空间，就扩容到了300G，结果仍然不够。
不过重启milvus的docker容器后就释放出来了，然后再慢慢涨上去

**Reply by LoveEachDay** [reply]
> 占用300G是因为centos-root只有300G。 原来是100G，满了后以为数据量大确实需要大空间，就扩容到了300G，结果仍然不够。 不过重启milvus的docker容器后就释放出来了，然后再慢慢涨上去

You use `ncdu` to show directory disk usage recursively.
